From b1eaffbbad39bde3ad989d949f9ce93c19a077d8 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:01:02 -0400
Subject: [fs] btrfs: send: remove BUG_ON from name_cache_delete

Message-id: <1407179109-13714-176-git-send-email-sandeen@redhat.com>
Patchwork-id: 87106
O-Subject: [RHEL7.1 PATCH 175/422] btrfs: send: remove BUG_ON from name_cache_delete
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: David Sterba <dsterba@suse.cz>

commit 57fb8910c24004ec924103c9a8c8542119f7629a
Author: David Sterba <dsterba@suse.cz>
Date:   Mon Feb 3 19:24:40 2014 +0100

    btrfs: send: remove BUG_ON from name_cache_delete

    If cleaning the name cache fails, we could try to proceed at the cost of
    some memory leak. This is not expected to happen often.

    Signed-off-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/send.c b/fs/btrfs/send.c
index d3ed9df..bef7ba6 100644
--- a/fs/btrfs/send.c
+++ b/fs/btrfs/send.c
@@ -1890,13 +1890,20 @@ static void name_cache_delete(struct send_ctx *sctx,
 
  nce_head = radix_tree_lookup(&sctx->name_cache,
    (unsigned long)nce->ino);
- BUG_ON(!nce_head);
+ if (!nce_head) {
+  btrfs_err(sctx->send_root->fs_info,
+       "name_cache_delete lookup failed ino %llu cache size %d, leaking memory",
+   nce->ino, sctx->name_cache_size);
+ }
 
  list_del(&nce->radix_list);
  list_del(&nce->list);
  sctx->name_cache_size--;
 
- if (list_empty(nce_head)) {
+ /*
+  * We may not get to the final release of nce_head if the lookup fails
+  */
+ if (nce_head && list_empty(nce_head)) {
   radix_tree_delete(&sctx->name_cache, (unsigned long)nce->ino);
   kfree(nce_head);
  }
-- 
1.7.1